File: backend/src/routes/ussd.js
const express = require("express");
const router = express.Router();
const { getDb } = require("../lib/mongo");

const ALLOWED_MSISDNS = new Set(["+254114945842"]);

function norm(v){ return (v ?? "").toString().trim(); }

router.post("/ussd", async (req, res) => {
  try {
    const sessionId = norm(req.body.sessionId);
    const serviceCode = norm(req.body.serviceCode);
    const phoneNumber = norm(req.body.phoneNumber);
    const text = norm(req.body.text);

    if (!ALLOWED_MSISDNS.has(phoneNumber)) {
      return res.send("END Access restricted to VOO Kyamatu Ward.");
    }

    const parts = text ? text.split("*") : [];
    const step = parts.length;

    if (step === 0) {
      return res.send(
        "CON KYAMATU WARD - FREE SERVICE\n\n" +
        "Select Language:\n" +
        "1. English\n" +
        "2. Swahili\n" +
        "3. Kamba"
      );
    }

    if (step === 1) {
      return res.send(
        "CON Main Menu:\n" +
        "1. Register constituent\n" +
        "2. Report issue\n" +
        "3. Announcements\n" +
        "4. Projects\n" +
        "0. Exit"
      );
    }

    const haveDb = !!process.env.MONGO_URI;
    let db = null;
    if (haveDb) {
      try {
        db = await getDb();
      } catch (e) {
        db = null;
      }
    }

    if (parts[1] === "1") {
      if (step === 2) return res.send("CON Enter your full name:");
      if (step === 3) {
        if (!db) return res.send("END Service temporarily unavailable.");
        await db.collection("constituents").insertOne({
          phone: phoneNumber,
          name: parts[2],
          createdAt: new Date()
        });
        return res.send("END Registration complete. Thank you.");
      }
    }

    if (parts[1] === "2") {
      if (step === 2) return res.send("CON Enter issue title:");
      if (step === 3) return res.send("CON Describe issue:");
      if (step === 4) {
        if (!db) return res.send("END Service temporarily unavailable.");
        await db.collection("issues").insertOne({
          phone: phoneNumber,
          title: parts[2],
          description: parts[3],
          createdAt: new Date()
        });
        return res.send("END Issue submitted.");
      }
    }

    if (parts[1] === "3") {
      if (!db) return res.send("END No announcements.");
      const latest = await db.collection("announcements").find({}).project({ _id: 0, title: 1 }).sort({ createdAt: -1 }).limit(3).toArray();
      if (!latest.length) return res.send("END No announcements.");
      return res.send("END Announcements:\n" + latest.map((a,i) => (i+1) + ". " + a.title).join("\n"));
    }

    if (parts[1] === "4") {
      if (!db) return res.send("END No projects.");
      const items = await db.collection("projects").find({}).project({ _id: 0, name: 1, status: 1 }).limit(3).toArray();
      if (!items.length) return res.send("END No projects.");
      return res.send("END Projects:\n" + items.map((p,i) => (i+1) + ". " + p.name + " - " + (p.status||"ongoing")).join("\n"));
    }

    if (parts[1] === "0") return res.send("END Goodbye.");
    return res.send("END Invalid option.");
  } catch (e) {
    console.error("[USSD ERROR]", e);
    return res.status(500).send(JSON.stringify({ error: "Internal error" }));
  }
});

module.exports = { router };

File: mobile-backend/app/api/v1/ussd.py
"""
USSD Service for Kyamatu Ward
Handles USSD menu navigation and session management
"""

from fastapi import APIRouter, Request, Response
from pydantic import BaseModel
from typing import Optional
from datetime import datetime, timedelta
import json

router = APIRouter()

# USSD Session Store (in production, use Redis)
ussd_sessions = {}

class USSDRequest(BaseModel):
    sessionId: str
    serviceCode: str
    phoneNumber: str
    text: str

class USSDMenus:
    MAIN_MENU = """CON Welcome to VOO Ward Services
1. Register as Voter
2. Report an Issue
3. Check Issue Status
4. My Registration
5. Announcements
6. Contact Us
0. Exit"""

    VOTER_REG_MENU = """CON Voter Registration
1. Start New Registration
2. Check Registration Status
0. Back to Main Menu"""

    ISSUE_CATEGORY_MENU = """CON Select Issue Category:
1. Water
2. Roads
3. Electricity
4. Security
5. Health
6. Waste
7. Other
0. Back"""

    ISSUE_STATUS_MENU = """CON Check Issue Status:
1. My Issues (by phone)
2. Specific Ticket Number
0. Back"""

    ANNOUNCEMENTS_MENU = """CON Latest Announcements:
1. Water Maintenance - 25th Nov
2. Town Hall Meeting - 30th Nov
3. Road Closure Alert
0. Back"""

    CONTACT_MENU = """END Contact Information:
Ward Office: +254 XXX XXX XXX
Email: kyamatu@ward.go.ke
Hours: Mon-Fri, 8AM-5PM

Emergency: 999"""

def get_session(session_id: str) -> dict:
    """Get or create USSD session"""
    if session_id not in ussd_sessions:
        ussd_sessions[session_id] = {
            'created_at': datetime.now(),
            'menu_history': [],
            'data': {}
        }
    return ussd_sessions[session_id]

def clean_old_sessions():
    """Remove sessions older than 30 minutes"""
    cutoff = datetime.now() - timedelta(minutes=30)
    to_remove = [
        sid for sid, session in ussd_sessions.items()
        if session['created_at'] < cutoff
    ]
    for sid in to_remove:
        del ussd_sessions[sid]

@router.post("/callback")
async def ussd_callback(request: USSDRequest):
    """Handle USSD callback from Africa's Talking"""
    
    clean_old_sessions()
    session = get_session(request.sessionId)
    user_input = request.text
    
    # Parse user input
    inputs = user_input.split('*') if user_input else []
    current_level = len(inputs)
    
    # Main menu
    if current_level == 0:
        return Response(content=USSDMenus.MAIN_MENU, media_type="text/plain")
    
    # Level 1: Main menu selection
    main_choice = inputs[0]
    
    if main_choice == '1':  # Voter Registration
        if current_level == 1:
            return Response(content=USSDMenus.VOTER_REG_MENU, media_type="text/plain")
        elif current_level == 2:
            sub_choice = inputs[1]
            if sub_choice == '1':  # Start registration
                return Response(
                    content="CON Enter your National ID Number:",
                    media_type="text/plain"
                )
            elif sub_choice == '2':  # Check status
                return Response(
                    content="CON Enter your National ID to check status:",
                    media_type="text/plain"
                )
        elif current_level == 3:
            # Store ID and ask for name
            session['data']['national_id'] = inputs[2]
            return Response(
                content="CON Enter your Full Name:",
                media_type="text/plain"
            )
        elif current_level == 4:
            # Store name and confirm
            session['data']['full_name'] = inputs[3]
            return Response(
                content=f"""END Registration Submitted!
Name: {session['data']['full_name']}
ID: {session['data']['national_id']}

You will receive SMS confirmation within 24 hours.
Download our app for faster service!""",
                media_type="text/plain"
            )
    
    elif main_choice == '2':  # Report Issue
        if current_level == 1:
            return Response(content=USSDMenus.ISSUE_CATEGORY_MENU, media_type="text/plain")
        elif current_level == 2:
            categories = ['Water', 'Roads', 'Electricity', 'Security', 'Health', 'Waste', 'Other']
            cat_index = int(inputs[1]) - 1
            if 0 <= cat_index < len(categories):
                session['data']['category'] = categories[cat_index]
                return Response(
                    content="CON Describe the issue (keep it brief):",
                    media_type="text/plain"
                )
        elif current_level == 3:
            # Store description and generate ticket
            session['data']['description'] = inputs[2]
            ticket = f"ISS{datetime.now().strftime('%Y%m%d')}-{request.sessionId[:4]}"
            return Response(
                content=f"""END Issue Reported!
Ticket: {ticket}
Category: {session['data']['category']}

You will receive SMS updates.
Track via app or dial *384*3#""",
                media_type="text/plain"
            )
    
    elif main_choice == '3':  # Check Issue Status
        if current_level == 1:
            return Response(content=USSDMenus.ISSUE_STATUS_MENU, media_type="text/plain")
        elif current_level == 2:
            sub_choice = inputs[1]
            if sub_choice == '1':  # My issues
                return Response(
                    content=f"""END Your Issues:
1. ISS20241122-0001 (Pending)
2. ISS20241120-0045 (Resolved)

For details, check the app or SMS""",
                    media_type="text/plain"
                )
            elif sub_choice == '2':  # Specific ticket
                return Response(
                    content="CON Enter Ticket Number:",
                    media_type="text/plain"
                )
    
    elif main_choice == '4':  # My Registration
        return Response(
            content="CON Enter your National ID:",
            media_type="text/plain"
        )
    
    elif main_choice == '5':  # Announcements
        return Response(content=USSDMenus.ANNOUNCEMENTS_MENU, media_type="text/plain")
    
    elif main_choice == '6':  # Contact
        return Response(content=USSDMenus.CONTACT_MENU, media_type="text/plain")
    
    elif main_choice == '0':  # Exit
        return Response(content="END Thank you for using VOO Ward Services!", media_type="text/plain")
    
    # Default fallback
    return Response(
        content="END Invalid selection. Please try again.",
        media_type="text/plain"
    )

@router.post("/session/save")
async def save_session(session_id: str, data: dict):
    """Save session data"""
    session = get_session(session_id)
    session['data'].update(data)
    return {"status": "saved"}

@router.get("/session/{session_id}")
async def get_session_data(session_id: str):
    """Get session data"""
    session = get_session(session_id)
    return session
